Change getAuthorities to reduce number of executed SQL-Statements

tack-sap opened the pull request:
Any time a user is retrieved using this class, the getAuthorities method is executed to retrieve all of the user's direct and indirect group memberships recursively. 
As mentioned in #463 by my colleague @mwdb the number of SQL queries used to do this grows with the number of scopes assigned.
As the methods are called very frequently, this highly increases the load on the database. With this PR the number of SQL queries is reduced. According to our performance tests this highly increases the performance of the database.

fhanik commented:
Thank you for the PR

It appears that we have a coverage decrease here. Meaning, not all lines of code added were exercised by the tests. We prefer code changes that are test driven. https://en.wikipedia.org/wiki/Test-driven_development

It may take longer to merge a PR, when we have to write supplementary tests or try to figure out what lines are not tested in this feature.

tack-sap commented:
Hi Filip,

With the added test all lines of the methods we changed are covered by tests (according to our tooling).

When we tried to cover all lines, we found, that the return in the catch-block was never executed by tests. The catch-block itself is taken from the original coding, but is empty there. We added the return because our guidelines do not allow empty catch-blocks.
The catch-block is never executed, because queryForList does not throw this exception on an empty result, but returns an empty list (which is why the catch was never used in the previous coding as well).
As there is no other catch, we removed then the try-catch from the method.

tack-sap commented:
Is there a reason why this PR is still "unscheduled" after such a long time?
All checks have been passed, the coverage report says "coverage increased", there a no conflicts and I got no feedback left, that some further action is required (as the feedback I got was already addressed with the second commit)..
Is there anything I can do for this PR to get processed?

sreetummidi commented:
We will get this scheduled for UAA 4.0

fhanik commented:
@tack-sap Thank you for your contribution. It is running through.

I corrected your test. In a Test Driven Development process, the test should fail before a change is applied, and succeed after. Your test succeeded before and after, thus the test does not detect a change in the code base.

I have modified your test - there is a neat trick using the `Mockito.spy` proxy

https://github.com/cloudfoundry/uaa/commit/eb0a970283cb8bf0ebf0d0b97308c73ca816ca5d

fhanik commented:
Merged via https://github.com/cloudfoundry/uaa/commit/d4a9baae31f97238cc0e02154e55816ed3f3854c
